server/main_test.go

```diff
@@ -0,0 +1,218 @@
+	// Ignore any locally defined SiteURL as we intend to host our own.
+	os.Unsetenv("MM_SERVICESETTINGS_SITEURL")
+	os.Unsetenv("MM_SERVICESETTINGS_LISTENADDRESS")
+
+	tmpDir, err := os.MkdirTemp("", "msteams")
+	if err != nil {
+		return err
+	}
+	mt.Cleanup(func() {
+		os.RemoveAll(tmpDir)
+	})
+
+	// Execute from the temporary directory to avoid polluting the developer's working
+	// directory and simplify cleanup.
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		return err
+	}
+
+	// Setup a custom MM_LOCALSOCKETPATH.
+	os.Setenv("MM_LOCALSOCKETPATH", path.Join(tmpDir, "mattermost_local.socket"))
+
+	// Create a test memory store and modify configuration appropriately
+	configStore := config.NewTestMemoryStore()
+	config := configStore.Get()
+	config.PluginSettings.Directory = model.NewString(path.Join(tmpDir, "plugins"))
+	config.PluginSettings.ClientDirectory = model.NewString(path.Join(tmpDir, "client"))
+	config.ServiceSettings.ListenAddress = model.NewString("localhost:0")
+	config.TeamSettings.MaxUsersPerTeam = model.NewInt(10000)
+	config.LocalizationSettings.SetDefaults()
+	config.SqlSettings = *storetest.MakeSqlSettings("postgres", false)
+	config.ServiceSettings.SiteURL = model.NewString("http://example.com/")
+	config.LogSettings.EnableConsole = model.NewBool(true)
+	config.LogSettings.EnableFile = model.NewBool(false)
+	config.LogSettings.ConsoleLevel = model.NewString("DEBUG")
+	config.ServiceSettings.EnableLocalMode = model.NewBool(true)
+	config.ServiceSettings.LocalModeSocketLocation = model.NewString(path.Join(tmpDir, "mattermost_local.socket"))
+	config.ServiceSettings.EnableDeveloper = model.NewBool(true)
+	config.ServiceSettings.EnableTesting = model.NewBool(true)
+	config.FileSettings.Directory = model.NewString(path.Join(tmpDir, "data"))
+
+	_, _, err = configStore.Set(config)
+	if err != nil {
+		return err
+	}
+
+	// Create a logger to override
+	testLogger, err := mlog.NewLogger()
+	if err != nil {
+		return err
+	}
+	testLogger.LockConfiguration()
+
+	// Initialize the server with app and api4 interfaces.
+	options := []app.Option{
+		app.ConfigStore(configStore),
+	}
+
+	server, err := app.NewServer(options...)
+	if err != nil {
+		return err
+	}
+
+	_, err = api4.Init(server)
+	if err != nil {
+		return err
+	}
+
+	err = server.Start()
+	if err != nil {
+		return err
+	}
+	mt.Cleanup(func() {
+		server.Shutdown()
+	})
+
+	ap := app.New(app.ServerConnector(server.Channels()))
+
+	// Setup the first user immediately.
+	username := model.NewId()
+	user := &model.User{
+		Email:         fmt.Sprintf("%s@example.com", username),
+		Username:      username,
+		Password:      "password",
+		EmailVerified: true,
+	}
+
+	_, appErr := ap.CreateUser(request.EmptyContext(testLogger), user)
+	if appErr != nil {
+		return appErr
+	}
+
+	// Export the site url for later use in tests.
+	os.Setenv("MM_SERVICESETTINGS_SITEURL", fmt.Sprintf("http://localhost:%v", server.ListenAddr.Port))
+
+	return nil
+}
```

Comment:
In the long run, I think we'll want both options, as I agree running against a particular server version is super powerful. But for now I think the defaults of the `go.mod` version is way more powerful, since with a simple `go mod replace`, I can make server-side changes and test them immediately with the plugin (e.g. think shared channels iteration). It's way more painful to force a Docker image.

```sh
go mod edit -replace=github.com/mattermost/mattermost/server/public=../../platform/server/public
go mod edit -replace=github.com/mattermost/mattermost/server/v8=../../platform/server
```

Reply:
One downside of using `go.mod` is that any breaking changes in `public` (and there have been ones) will hinder using an older version for tests.

This is surely not a blocker for this PR, but I wonder if we would benefit from decoupling the MM version the plugin is compiled against from the MM version the plugin runs against. (This is also how a "real" manual test would work!)

Reply:
Agreed, and this is where the Docker approach could work well. It should be an easy extension to what we have now.